src/platform_impl/linux/wayland/event_loop.rs

```diff
@@ -69,6 +74,68 @@ impl<T> WindowEventsSink<T> {
     }
 }
 
+pub struct CursorManager {
+    pointer_constraints_proxy: Arc<Mutex<Option<ZwpPointerConstraintsV1>>>,
+    pointers: Vec<wl_pointer::WlPointer>,
+    locked_pointers: Vec<ZwpLockedPointerV1>,
+    cursor_visible: Arc<Mutex<bool>>,
+}
+
+impl CursorManager {
+    fn new(constraints: Arc<Mutex<Option<ZwpPointerConstraintsV1>>>) -> CursorManager {
+        CursorManager {
+            pointer_constraints_proxy: constraints,
+            pointers: Vec::new(),
+            locked_pointers: Vec::new(),
+            cursor_visible: Arc::new(Mutex::new(true)),
+        }
+    }
+
+    fn register_pointer(&mut self, pointer: wl_pointer::WlPointer) {
+        self.pointers.push(pointer);
+    }
+
+    pub fn set_cursor_visible(&mut self, visible: bool) {
+        // If we want to hide the cursor do it immediately. There is a known
+        // limitation that the cursor will not become visible again until it
+        // enters a new surface
```

Comment:
Please, explain a bit more this "known limitation", especially where does it come from? Is it an inherent limitation from the wayland protocol? Is it a `TODO` that should be improved in the future?

It seems that the rest of your code makes no attempt to set the pointer visible when requested, so I'd really prefer this to be explained.

Reply:
What I meant was that it is a known limitation in my winit implementation because I was too lazy to figure out how to obtain the default cursor surface which I would need to pass into set_cursor call. 

I figured it would be good enough for now because wayland will revert to the default cursor on the next PtrEvent::Enter. 

It's definitively a `TODO` though. Should I change the comment to me more clear? Or do we absolute need the feature of being able to make the cursor visible again right away? Maybe obtaining the default cursor WlSurface is trivial, but I am having a hard time finding documentation on these thing.

Reply:
I'm not sure if I'm correct here, but I was doing kind of the same thing when implementing this `set_visible`, expect I was using `AutoThemer` and `AutoPointer` from SCTK to change cursor, which is quite easy to manage cursor and overcome this limitation iirc. Since my PR is kind of WIP and private, and your one was earlier you can try adopting `AutoPointer`/`AutoThemer` approach.

Reply:
fwiw without setting cursor visible right away it's quite buggy feature and will need immediate fix in my opinion (At least from alacritty usage point of view).

Reply:
Indeed, `AutoThemer` provided by SCTK should make the process of making the cursor visible back again easily, and also change the icon to other things (grabbing cursor, loading icon, ...).

I'm not opposed to first merging this as is, and adding AutoThemer support in a future PR, but in that case please add a clear `TODO` comment explaining what remains to be done, so that we won't forget about this.

Reply:
I updated the comments in the code and also added documentation about this in the public API.

Reply:
It seems I have gotten it to work using ThemeManager and ThemedPointer. Currently I am holding a vector of both the registered WlPointers and a vector of ThemedPointers which does not seem optimal to me. There reason being that I don't want the entire thing to stop working if for some reason the WlShm or ThemeManager fails to initialize. Is this a valid concern? Or can I assume that this will always work?

Reply:
`WlShm` cannot fail to initialize, or it is a bug from the server, so we don't need to worry about this.

wrt to `ThemedPointer`, this is what [`AutoThemer`](https://docs.rs/smithay-client-toolkit/0.6.4/smithay_client_toolkit/pointer/enum.AutoThemer.html) is for: it gracefully handles a failure to initialize a `ThemeManager`. And in that case, you only need to keep the `AutoPointer`s, as they give you access to the `WlPointer`s via deref.

Reply:
I've switched to using `AutoThemer` and `AutoPointer`. Now the `CursorManager` only stores a vector of `AutoPointer` objects which I think is better.

I encountered some difficulties implementing this. The first problem was that the second call to `sync_roundtrip` inside `from_display_with_cb` triggered the registering of the pointer, but I was only initializing the `AutoThemer` after creating the environment. This caused caused panic when registering the pointer in `CursorManager` because the `AutoThemer` was not set when registering the pointer.

To work around that I moved the initialization of the `AutoThemer` into the `Environment` callback. Then there was the problem that you need both shm and compositor to initialize `AutoThemer`. So I had to set up the compositor interface as well and only trigger the code to initialize the `AutoThemer` once both the compositor and the shm interface was initialized.

Could it be a good idea to have some way for the caller to run some code after all the global interface callbacks have been run, but before the second call to `sync_roundtrip` ? Or maybe I am just using this API completely wrong?

Either way, I think all the mentioned issues has been resolved in this pull request.

Reply:
This is likely that SCTK's API has room for improvement. I've actually opened an issue to gather feedback there: https://github.com/Smithay/client-toolkit/issues/77.